app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -30,39 +35,42 @@ def logout
 
   private
 
-  # TODO: get this from a common place like ApplicationController
-  def user_signed_in?
-    session[:user_id].present? && User.find_by(id: session[:user_id])
+  def auth_info
+    @auth_info ||= request.env['omniauth.auth']['info']
   end
-  helper_method :user_signed_in?
 
-  def google_callback
-    auth = request.env['omniauth.auth']
-    auth_info = auth['info']
-    email     = auth_info['email']
-
-    if email.blank?
-      redirect_to '/auth/google_oauth2?auth_type=rerequest&scope=email'
-      return
-    end
+  def saml_response_attributes
+    @saml_response_attributes ||= request.env['omniauth.auth'].extra.response_object.attributes
+  end
 
-    user = User.find_or_initialize_by(email: email) do |u|
-      u.email       = email
+  def user
+    @user ||= User.find_or_initialize_by(email: auth_info['email']) do |u|
+      u.email       = auth_info['email']
       u.first_name  = auth_info['first_name']
       u.last_name   = auth_info['last_name']
     end
+  end
 
-    # Update the user's info
-    user.google_token = auth['credentials']['token']
-    user.photo        = auth_info['image']
-    user.locale       = request.env.dig('omniauth.auth', 'extra', 'raw_info', 'locale')
+  def set_user_photo
+    user.photo = auth_info['image'] if auth_info['image']
+  end
 
-    user.save!
+  def set_user_metadata
+    user.metadata = saml_response_attributes.attributes.keys.each_with_object({}) do |key, metadata|
+      # Ignore default fields we already capture in user or don't care about
+      next if ['first_name', 'last_name', 'email', 'fingerprint'].include?(key)
+      metadata[key] = saml_response_attributes[key]
+    end.to_json
+  end
 
-    session[:user_id] = user.id
+  # TODO: get this from a common place like ApplicationController
+  def user_signed_in?
+    session[:user_id].present? && User.find_by(id: session[:user_id])
+  end
+  helper_method :user_signed_in?
 
-    redirect_to params[:return_to] || :portal
-  rescue ActiveRecord::RecordInvalid => e
-    render text: e.message, status: :forbidden
+  def verify_callback_contents
+    raise 'Authentication provider must provide email, first_name, and last_name parameters' unless (auth_info.keys & ['email', 'first_name', 'last_name']).size == 3
```

Comment:
line is pretty long, maybe turn into an early return guard or multiline the unless?